tests/token/erc721/test_ERC721_SafeMintable_mock.py

```diff
@@ -0,0 +1,195 @@
+    erc721, account, _, erc721_holder, _ = erc721_factory
+
+    await signer.send_transaction(
+        account, erc721.contract_address, 'safeMint', [
+            erc721_holder.contract_address,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ]
+    )
+
+    # check balance
+    execution_info = await erc721.balanceOf(erc721_holder.contract_address).call()
+    assert execution_info.result == (to_uint(1),)
+
+    # check owner
+    execution_info = await erc721.ownerOf(TOKEN).call()
+    assert execution_info.result == (erc721_holder.contract_address,)
+
+
+@pytest.mark.asyncio
+async def test_safeMint_to_account(erc721_factory):
+    erc721, account, recipient, _, _ = erc721_factory
+
+    await signer.send_transaction(
+        account, erc721.contract_address, 'safeMint', [
+            recipient.contract_address,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ]
+    )
+
+    # check balance
+    execution_info = await erc721.balanceOf(recipient.contract_address).call()
+    assert execution_info.result == (to_uint(1),)
+
+    # check owner
+    execution_info = await erc721.ownerOf(TOKEN).call()
+    assert execution_info.result == (recipient.contract_address,)
+
+
+@pytest.mark.asyncio
+async def test_safeMint_to_zero_address(erc721_factory):
+    erc721, account, _, _, _ = erc721_factory
+
+    # to zero address should be rejected
+    await assert_revert(signer.send_transaction(
+        account, erc721.contract_address, 'safeMint', [
+            ZERO_ADDRESS,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ]
+    ))
+
+
+@pytest.mark.asyncio
+async def test_safeMint_from_zero_address(erc721_factory):
+    erc721, _, _, erc721_holder, _ = erc721_factory
+
+    # Caller address is `0` when not using an account contract
+    await assert_revert(
+        erc721.safeMint(
+            erc721_holder.contract_address,
+            TOKEN,
+            DATA
+        ).invoke()
+    )
+
+
+@pytest.mark.asyncio
+async def test_safeMint_from_not_owner(erc721_factory):
+    erc721, _, other, erc721_holder, _ = erc721_factory
+
+    await assert_revert(signer.send_transaction(
+        other, erc721.contract_address, 'safeMint', [
+            erc721_holder.contract_address,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ])
+    )
+
+
+@pytest.mark.asyncio
+async def test_safeMint_to_unsupported_contract(erc721_factory):
+    erc721, account, _, _, unsupported = erc721_factory
+
+    try:
+        await signer.send_transaction(
+            account, erc721.contract_address, 'safeMint', [
+                unsupported.contract_address,
+                *TOKEN,
+                len(DATA),
+                *DATA
+            ]
+        )
+        assert False
+    except StarkException as err:
+        _, error = err.args
+        assert error['code'] == StarknetErrorCode.ENTRY_POINT_NOT_FOUND_IN_CONTRACT
```

Comment:
can't this be replaced with `assert_revert`?

Reply:
Not as it is right now. `assert_revert` checks for the specific error code `TRANSACTION_FAILED`, but I'll take out the error code check to make it more generalized: 
```
async def assert_revert(fun, reverted_with=None):
   try:
       await fun
       assert False
   except StarkException as err:
       _, error = err.args
       if reverted_with is not None:
           assert reverted_with in error['message']
```